cmd/repo-brancher/main.go

```diff
@@ -0,0 +1,167 @@
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/openshift/ci-operator-prowgen/pkg/config"
+	"github.com/openshift/ci-operator-prowgen/pkg/promotion"
+	"github.com/openshift/ci-operator/pkg/api"
+)
+
+type options struct {
+	promotion.Options
+	gitDir      string
+	username    string
+	password    string
+	fastForward bool
+}
+
+func (o *options) Validate() error {
+	if err := o.Options.Validate(); err != nil {
+		return err
+	}
+	if o.Confirm {
+		if o.username == "" {
+			return errors.New("--username is required with --confirm")
+		}
+		if o.password == "" {
+			return errors.New("--password is required with --confirm")
+		}
+	}
+	return nil
+}
+
+func gatherOptions() options {
+	o := options{}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	fs.StringVar(&o.gitDir, "git-dir", "", "Optional dir to do git operations in. If unset, temp dir will be used.")
+	fs.StringVar(&o.username, "username", "", "Username to use when pushing to GitHub.")
+	fs.StringVar(&o.password, "password", "", "Password to use when pushing to GitHub.")
+	fs.BoolVar(&o.fastForward, "fast-forward", false, "Attempt to fast-forward future branches if they already exist.")
+	o.Bind(fs)
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		logrus.WithError(err).Fatal("could not parse input")
+	}
+	return o
+}
+
+func main() {
+	o := gatherOptions()
+	if err := o.Validate(); err != nil {
+		logrus.Fatalf("Invalid options: %v", err)
+	}
+
+	gitDir := o.gitDir
+	if gitDir == "" {
+		tempDir, err := ioutil.TempDir("", "")
+		if err != nil {
+			logrus.WithError(err).Fatal("Could not create temp dir for git operations")
+		}
+		defer func() {
+			if err := os.Remove(tempDir); err != nil {
+				logrus.WithError(err).Fatal("Could not clean up temp dir for git operations")
+			}
+		}()
+		gitDir = tempDir
+	}
+
+	if err := config.OperateOnCIOperatorConfigDir(o.ConfigDir, func(configuration *api.ReleaseBuildConfiguration, repoInfo *config.FilePathElements) error {
+		logger := config.LoggerForInfo(*repoInfo)
```

Comment:
This was a bit heavy to review, that anonymous method is quite long and could be broken down so some of the details are abstracted away.